mariamihaelafainarea/assignment3/Assignment3

```diff
@@ -0,0 +1,171 @@
+            int newX = x + dx[i];
+            int newY = y + dy[i];
+            if (newX >= 0 && newX < n) {
+                if (newY >= 0 && newY < m) {
+                    if(visited[newX][newY] == false) {
+                        visited[newX][newY] = true;
+                        StringBuilder hopa = start.append(grid[newX][newY]);
+                        if (dictionary.isPrefix(hopa.toString())){
+                            if (dictionary.isWord(hopa.toString())) {
+                                theList.add(hopa.toString());
+                            }
+                            DFSonGrid(theList, hopa, newX, newY,visited);
+                        }
+                        //restart the state after a deep search ! 
+                        start.deleteCharAt(hopa.length()-1);
+                        visited[newX][newY] = false;
+                    }
+                }
+            }
+        }
+        
+    }
+    //A dictionary based on a trie ! 
+    //It has a better complexity than
+    //a hashmap !
+    class MyDictionary{
+        private class TrieNode{
+            private HashMap<Character,TrieNode> children;
+            private boolean endOfWord;
+            public TrieNode() {
+                children = new HashMap<>();
+                endOfWord = false;
+            }
+            
+        }
+        private final TrieNode root;
+        //the idea is very simple
+        //you start from root
+        //you go deep down in the trie 
+        //with the current node
+        //if theChildren (the character that need to be added)
+        //exists you only pass it and go deep down
+        public void insert(String word) {
+            TrieNode current = root;
+            for(int i = 0; i < word.length();i++) {
+                TrieNode theChild = current.children.get(word.charAt(i));
+                if (theChild == null) {  
+                    theChild = new TrieNode();
+                    current.children.put(word.charAt(i), theChild);
+                }
+                current = theChild;
+            }
+            //after you add the last node and move your
+            //current there
+            //you mark it as endOfWord
+            current.endOfWord = true;
+            
+        }
+        public boolean isGeneralHelper(String action,TrieNode current,String word,int index) {
+            if(action.equals("word")) {
+                if (index == word.length()) {
+                    return current.endOfWord;
+                }
+            } else {
+                if (index == word.length()) {
+                    if(current == null) {
+                        return false;
+                    }
+                return true;
+                }
+            }
+            
+            TrieNode theChildren = current.children.get(word.charAt(index));
+            if (theChildren == null) {
+               return false;
+            }
+            index ++;
```

Comment:
You can increment index in the next line when calling the recursive call. Just make sure you increment it before it is used: ++index instead of index++.

Reply:
done